### PR TITLE
Automatic cancelation

### PR DESCRIPTION
**Pull request by MStodolkiewicz**
Added redis, sidekiq

Added job for automatic reservation cancelation

### CODE REVIEW

File: app/jobs/cancel_reservations_job.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    end
+  end
+
+  private
+
+  def not_confirmed_on_time?(reservation)
+    reservation.seance.start_time <= 30.minutes.from_now
+  end
+
+  def reservations
```
Comment (w0st) [inline app/jobs/cancel_reservations_job.rb]:
```suggestion
  def pending_reservations
```

it even brings more info to readers :wink:

File: app/jobs/cancel_reservations_job.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  def perform
+    reservations.each do |reservation|
+      cancel_reservation(reservation) if not_confirmed_on_time?(reservation)
+    end
+  end
+
+  private
+
+  def not_confirmed_on_time?(reservation)
+    reservation.seance.start_time <= 30.minutes.from_now
```
Comment (w0st) [inline app/jobs/cancel_reservations_job.rb]: Are you sure that this condition is valid? :wink: 

It looks like the most crucial thing in this PR, why don't you want to cover it by unit test?

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
+require 'sidekiq/web'
+require 'sidekiq/cron/web'
+
 Rails.application.routes.draw do
+  mount Sidekiq::Web => '/sidekiq'
```
Comment (w0st) [inline config/routes.rb]: It's fine for educational purposes. Please notice that this endpoint shouldn't be publicly accessible in production (should require authentication & authorization) :wink:

**Review by w0st — COMMENTED** [review]
Consider my comments, please.

File: app/jobs/cancel_reservations_job.rb (hunk 1 above)
Comment (MStodolkiewicz) [reply app/jobs/cancel_reservations_job.rb]: True, changed that :D

File: app/jobs/cancel_reservations_job.rb (hunk 2 above)
Comment (MStodolkiewicz) [reply app/jobs/cancel_reservations_job.rb]: Yup, it's valid. I wrote 3 tests for start_time being less than, equal, and more than 30 minutes :D.

File: config/routes.rb (hunk 3 above)
Comment (MStodolkiewicz) [reply config/routes.rb]: I have found this solution on sidekiq github wiki. Would that be enough?

```
authenticate :user, lambda { |u| u.admin? } do
  mount Sidekiq::Web => '/sidekiq'
end
```

File: config/sidekiq.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    config.redis = { url: 'redis://127.0.0.1:6379' } 
+  end
+  Sidekiq.configure_client do |config|   
+    config.redis = { url: 'redis://127.0.0.1:6379' } 
+  end
+  
+  schedule_file = 'config/schedule.yml'
+  if File.exist?(schedule_file) && Sidekiq.server?
+    Sidekiq::Cron::Job.load_from_hash YAML.load_file(schedule_file)
+  end
```
Comment (rrothenberger) [inline config/sidekiq.rb]: This part should be inside `Sidekiq.configure_server` block - we want this to only load in sidekiq server(which is the thing we start with `bundle exec sidekiq`). And I believe it should be specified to be started only on startup - I can't link for specific part of code, but just above [Finding jobs](https://github.com/ondrejbartas/sidekiq-cron#finding-jobs) you will find example from repo how to load this to sidekiq-cron 🙂

Comment (rrothenberger) [reply config/sidekiq.rb]: And could you fix the indentation in this file, please? 🙏

File: spec/requests/seance_spec.rb [hunk 5]
```diff
@@ -48,11 +48,13 @@
...
   describe "POST /seances" do
     subject(:request) { post('/seances', params:) }
-    
+
     let(:movie) { create :movie }
     let(:hall) { create :hall }
 
-    let(:params) { { seance: attributes_for(:seance, start_time: DateTime.current, price:, movie_id: movie.id, hall_id: hall.id ) } }
+    let(:params) do
+      { seance: attributes_for(:seance, start_time: DateTime.current, price:, movie_id: movie.id, hall_id: hall.id) }
```
Comment (rrothenberger) [inline spec/requests/seance_spec.rb:56]: O - I didn't know about `attributes_for` of FactoryBot, you learn something new everyday 😄

File: app/jobs/cancel_reservations_job.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def pending_reservations
+    Reservation.where(status: :reserved)
+  end
```
Comment (rrothenberger) [inline app/jobs/cancel_reservations_job.rb]: Why not query for reservations that have `status` reserved and screenings that start in 30 minutes or less?
```ruby
Reservation.joins(:seance).where(status: :reserved).where(seance: {start_time: ..30.minutes.from_now})
```

Something like that would probably work, it joins reservations with seances and then takes only those reservations for which seances starts in 30 minutes or less. That way you will take only those reservations that you should be cancelling, no all reservation in reserved states - after all, there will be much more of those.

File: spec/sidekiq/cancel_reservations_job_spec.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      end
+    end
+  end
+
+  describe 'perform_later' do
+    it 'enqueues job' do
+      ActiveJob::Base.queue_adapter = :test
+      expect { described_class.perform_later }.to have_enqueued_job
+    end
+  end
```
Comment (rrothenberger) [inline spec/sidekiq/cancel_reservations_job_spec.rb]: You don't really need to test that - you are basically testing the library here. `have_enqueued_job` is meant for places when you are actually calling a job - so not really applicable for cron jobs.

For example you might have: `ClearSeancesCacheJob` that will be scheduled in some controller, everytime somebody adds, changes, or removes a seance. Then when testing the controller you would use `have_enqueued_job` to check if this job is actually scheduled on that actions 😉

File: config/sidekiq.rb (hunk 4 above)
Comment (MStodolkiewicz) [reply config/sidekiq.rb]: My bad, saw this config and it worked so i didn't check the official documentation, should be more careful.

File: spec/sidekiq/cancel_reservations_job_spec.rb (hunk 7 above)
Comment (MStodolkiewicz) [reply spec/sidekiq/cancel_reservations_job_spec.rb]: We talked about this on Friday and it completely slipped my mind to delete it.

File: app/jobs/cancel_reservations_job.rb (hunk 6 above)
Comment (MStodolkiewicz) [reply app/jobs/cancel_reservations_job.rb]: I made the changes and I'm using this query now. True, it was unnecessary taking all thous reservations.

**Review by rrothenberger — APPROVED** [review]

**Pull request merged**